File: Data Structures & Algorithms/swim-in-rising-water/submission-18.py
class Solution:
    def swimInWater(self, grid: List[List[int]]) -> int:
        n = len(grid)
        minHeap = [(grid[0][0],0,0)]
        res = 0
        visit = [[False] * n for _ in range(n)]
        directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        while minHeap:
            t, r, c = heapq.heappop(minHeap)
            if visit[r][c]:
                continue
            if r == n - 1 and c == n - 1:
                return t
            visit[r][c] = True
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                if min(nr, nc) < 0 or max(nr, nc) >= n or visit[nr][nc]:
                    continue
                heapq.heappush(minHeap, (max(t, grid[nr][nc]), nr, nc))
        

File: Data Structures & Algorithms/trapping-rain-water/submission-18.cs
public class Solution {
    public int Trap(int[] height) {
        var length = height.Length;
        if (length == 0) {
            return 0;
        }
        var prefix = new int[length];
        prefix[0] = height[0];
        var suffix = new int[length];
        suffix[length - 1] = height[length - 1];

        for (int i = 1; i < length; i++) {
            prefix[i] = Math.Max(prefix[i - 1], height[i]);
        }

        for (int i = length - 2; i >= 0; i--) {
            suffix[i] = Math.Max(suffix[i + 1], height[i]);
        }

        var res = 0;
        for (int i = 0; i < length; i++) {
            res += Math.Min(prefix[i], suffix[i]) - height[i];
        }
        return res;
    }
}
